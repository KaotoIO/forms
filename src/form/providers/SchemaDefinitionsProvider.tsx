import { JSONSchema7, JSONSchema7Definition } from 'json-schema';
import { createContext, FunctionComponent, PropsWithChildren, useMemo } from 'react';

interface SchemaDefinitionsContextValue {
  definitions: JSONSchema7Definition;
  omitFields: string[];
}

export const SchemaDefinitionsContext = createContext<SchemaDefinitionsContextValue>({
  definitions: {},
  omitFields: [],
});

export const SchemaDefinitionsProvider: FunctionComponent<
  PropsWithChildren<{ schema: JSONSchema7; omitFields: string[] }>
> = ({ schema, omitFields, children }) => {
  const value = useMemo(() => {
    const definitions: SchemaDefinitionsContextValue['definitions'] = schema.definitions ?? {};

    return { definitions, omitFields };
  }, [omitFields, schema.definitions]);

  return <SchemaDefinitionsContext.Provider value={value}>{children}</SchemaDefinitionsContext.Provider>;
};
