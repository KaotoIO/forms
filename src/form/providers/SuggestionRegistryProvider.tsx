import { JSONSchema7 } from 'json-schema';
import { createContext, FunctionComponent, PropsWithChildren, useCallback, useContext, useMemo, useState } from 'react';
import { SuggestionProvider } from '../models/suggestions';

interface SuggestionContextApi {
  registerProvider: (provider: SuggestionProvider) => void;
  unregisterProvider: (id: string) => void;
}
interface SuggestionContext {
  getProviders: (propertyName: string, schema: JSONSchema7) => SuggestionProvider[];
}

export const SuggestionContextApi = createContext<SuggestionContextApi | undefined | null>(undefined);
export const SuggestionContext = createContext<SuggestionContext>({
  getProviders: () => [],
});

export const SuggestionRegistryProvider: FunctionComponent<PropsWithChildren> = ({ children }) => {
  const [providers, setProviders] = useState<SuggestionProvider[]>([]);

  const registerProvider = useCallback((provider: SuggestionProvider) => {
    setProviders((prevProviders) => {
      if (prevProviders.some((p) => p.id === provider.id)) {
        console.warn(`Suggestion provider with ID "${provider.id}" already registered.`);
        return prevProviders;
      }
      return [...prevProviders, provider];
    });
  }, []);

  const unregisterProvider = useCallback((id: string) => {
    setProviders((prevProviders) => prevProviders.filter((p) => p.id !== id));
  }, []);

  const getProviders = useCallback(
    (propertyName: string, schema: JSONSchema7) => {
      if (!schema) return providers;
      return providers.filter((provider) => provider.appliesTo(propertyName, schema));
    },
    [providers],
  );

  const contextApi = useMemo(
    () => ({
      registerProvider,
      unregisterProvider,
    }),
    [registerProvider, unregisterProvider],
  );

  return (
    <SuggestionContextApi.Provider value={contextApi}>
      <SuggestionContext.Provider value={{ getProviders }}>{children}</SuggestionContext.Provider>
    </SuggestionContextApi.Provider>
  );
};

export const useSuggestionRegistry = () => {
  const context = useContext(SuggestionContextApi);
  if (context === undefined) {
    throw new Error('useSuggestionRegistry must be used within a SuggestionRegistryProvider');
  }
  return context;
};
