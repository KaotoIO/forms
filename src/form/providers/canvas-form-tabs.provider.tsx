import { FunctionComponent, PropsWithChildren, createContext, useState } from 'react';

type TabKeys = keyof typeof FormTabsModes;
export interface CanvasFormTabsContextResult {
  selectedTab: TabKeys;
  setSelectedTab: (key: TabKeys) => void;
}

export const CanvasFormTabsContext = createContext<CanvasFormTabsContextResult>({
  selectedTab: 'Required',
  setSelectedTab: () => void 0,
});

/**
 * Used for fetching and injecting the selected tab information from the canvas form
 */
export const CanvasFormTabsProvider: FunctionComponent<PropsWithChildren<{ tab?: TabKeys }>> = ({
  tab = 'Required',
  children,
}) => {
  const [selectedTab, setSelectedTab] = useState<TabKeys>(tab);

  return (
    <CanvasFormTabsContext.Provider
      value={{
        selectedTab,
        setSelectedTab,
      }}
    >
      {children}
    </CanvasFormTabsContext.Provider>
  );
};

export const FormTabsModes = {
  Required: 'Shows Required fields only',
  All: 'Shows All fields',
  Modified: 'Shows Modified fields only',
};
