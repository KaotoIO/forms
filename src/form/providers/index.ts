export * from './canvas-form-tabs.provider';
export * from './filtered-field.provider';
export * from './FormComponentFactoryProvider';
export * from './ModelProvider';
export * from './SchemaDefinitionsProvider';
export * from './SchemaProvider';
export * from './SuggestionRegistryProvider';
