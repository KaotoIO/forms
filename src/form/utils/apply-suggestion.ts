import { Suggestion } from '../models/suggestions';
import { getCursorWord } from './get-cursor-word';

export const applySuggestion = (
  suggestion: Suggestion,
  inputValue: string | number,
  cursorPosition?: number | null,
): { newValue: string; cursorPosition: number } => {
  const cursor = cursorPosition ?? 0;

  const { word, cursorStart, cursorEnd } = getCursorWord(inputValue, cursorPosition);

  const firstSection = inputValue.toString().substring(0, cursorStart);
  const lastSection = inputValue.toString().substring(cursorEnd);
  const newValue = firstSection + suggestion.value + lastSection;

  return {
    newValue,
    cursorPosition: cursorStart + suggestion.value.length,
  };
};
