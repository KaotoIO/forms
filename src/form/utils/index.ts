export * from './camel-case-to-space';
export * from './camel-random-id';
export * from './get-field-groups';
export * from './get-filtered-properties';
export * from './get-item-from-schema';
export * from './get-value';
export * from './is-defined';
export * from './is-raw-string';
export * from './resolve-schema-with-ref';
export * from './set-value';
