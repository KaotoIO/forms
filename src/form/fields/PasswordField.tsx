import { Button, TextInputGroup, TextInputGroupMain, TextInputGroupUtilities } from '@patternfly/react-core';
import { EyeIcon, EyeSlashIcon } from '@patternfly/react-icons';
import { FunctionComponent, useCallback, useContext, useRef, useState } from 'react';
import { useFieldValue } from '../hooks/field-value';
import { useSuggestions } from '../hooks/suggestions';
import { FieldProps } from '../models/typings';
import { SchemaContext } from '../providers/SchemaProvider';
import { isDefined, isRawString } from '../utils';
import { FieldActions } from './FieldActions';
import { FieldWrapper } from './FieldWrapper';

export const PasswordField: FunctionComponent<FieldProps> = ({ propName, required, onRemove: onRemoveProps }) => {
  const { schema } = useContext(SchemaContext);
  const [passwordHidden, setPasswordHidden] = useState<boolean>(true);
  const { value = '', errors, disabled, isRaw, onChange } = useFieldValue<string>(propName);

  const lastPropName = propName.split('.').pop();
  const clearButtonAriaLabel = isDefined(onRemoveProps) ? 'Remove' : `Clear ${lastPropName} field`;
  const toggleRawAriaLabel = `Toggle RAW wrap for ${lastPropName} field`;
  const inputRef = useRef<HTMLInputElement>(null);

  const onFieldChange = useCallback(
    (_event: unknown, value: string) => {
      onChange(value);
    },
    [onChange],
  );

  const onRemove = () => {
    if (isDefined(onRemoveProps)) {
      onRemoveProps(propName);
      return;
    }

    /** Clear field by removing its value */
    onChange(undefined as unknown as string);
    inputRef.current?.focus();
  };

  const toggleRawValueWrap = () => {
    if (typeof value !== 'string') return;

    const newValue = isRawString(value) ? value.substring(4, value.length - 1) : `RAW(${value})`;
    onChange(newValue);
  };

  const setValue = useCallback(
    (value: string | number) => {
      onFieldChange(null, value.toString());
    },
    [onFieldChange],
  );

  const suggestions = useSuggestions({
    propName,
    schema,
    inputRef,
    value,
    setValue,
  });

  const id = `${propName}-popover`;

  return (
    <FieldWrapper
      propName={propName}
      required={required}
      title={schema.title}
      type="secret"
      description={schema.description}
      defaultValue={schema.default?.toString()}
      errors={errors}
      isRaw={isRaw}
    >
      <TextInputGroup validated={errors ? 'error' : undefined} isDisabled={disabled}>
        <TextInputGroupMain
          type={passwordHidden ? 'password' : 'text'}
          role="textbox"
          id={propName}
          name={propName}
          placeholder={schema.default?.toString()}
          value={value}
          onChange={onFieldChange}
          aria-label={schema.title}
          aria-describedby={id}
          ref={inputRef}
        />

        {suggestions}

        <TextInputGroupUtilities>
          <Button
            variant="plain"
            data-testid={`${propName}__toggle-visibility`}
            aria-label={passwordHidden ? `Show ${lastPropName} value` : `Hide ${lastPropName} value`}
            onClick={() => {
              setPasswordHidden(!passwordHidden);
            }}
          >
            {passwordHidden ? <EyeIcon /> : <EyeSlashIcon />}
          </Button>

          <FieldActions
            propName={propName}
            clearAriaLabel={clearButtonAriaLabel}
            toggleRawAriaLabel={toggleRawAriaLabel}
            onRemove={onRemove}
            toggleRawValueWrap={toggleRawValueWrap}
          />
        </TextInputGroupUtilities>
      </TextInputGroup>
    </FieldWrapper>
  );
};
