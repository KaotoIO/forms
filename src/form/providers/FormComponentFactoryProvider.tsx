import { JSONSchema4 } from 'json-schema';
import { FunctionComponent, PropsWithChildren, useCallback } from 'react';
import { ArrayField } from '../fields/ArrayField/ArrayField';
import { BooleanField } from '../fields/BooleanField';
import { DisabledField } from '../fields/DisabledField';
import { EnumField } from '../fields/EnumField';
import { AllOfField } from '../fields/ObjectField/AllOfField';
import { ObjectField } from '../fields/ObjectField/ObjectField';
import { OneOfField } from '../fields/OneOfField/OneOfField';
import { PasswordField } from '../fields/PasswordField';
import { PropertiesField } from '../fields/PropertiesField/PropertiesField';
import { StringField } from '../fields/StringField';
import { TextAreaField } from '../fields/TextAreaField';
import { FieldProps } from '../models/typings';
import {
  FormComponentFactoryContext,
  FormComponentFactoryContextValue,
} from './context/form-component-factory-context';

export type CustomFieldsFactory = (schema: JSONSchema4) => FunctionComponent<FieldProps> | undefined;
interface IFormComponentFactoryProvider extends PropsWithChildren {
  /**
   * Factory function to override the Field selection algorithm.
   *
   * Should return a component receiving FieldProps as props, or undefined if no custom field is found.
   */
  customFieldsFactory?: CustomFieldsFactory;
}
export const FormComponentFactoryProvider: FunctionComponent<IFormComponentFactoryProvider> = ({
  customFieldsFactory,
  children,
}) => {
  const factory = useCallback<FormComponentFactoryContextValue>(
    (schema) => {
      const CustomField = customFieldsFactory?.(schema);
      if (CustomField) {
        return CustomField;
      }

      if (schema.format === 'password') {
        return PasswordField;
      } else if (schema.type === 'string' && schema.title && TEXT_AREA_PROPERTY_NAMES.includes(schema.title)) {
        return TextAreaField;
      } else if (schema.type === 'string' && Array.isArray(schema.enum)) {
        return EnumField;
      } else if (schema.type === 'object' && Object.keys(schema?.properties ?? {}).length === 0) {
        /*
         * If the object has no properties, we render a generic key-value pairs field
         * This is useful for langchain4j-tools consumer components or when configuring beans entities
         */
        return PropertiesField;
      }

      switch (schema.type) {
        case 'string':
        case 'number':
        case 'integer':
          return StringField;
        case 'boolean':
          return BooleanField;
        case 'object':
          return ObjectField;
        case 'array':
          return ArrayField;
      }

      if (Array.isArray(schema.oneOf)) {
        return OneOfField;
      } else if (Array.isArray(schema.allOf)) {
        return AllOfField;
      } else if (Array.isArray(schema.anyOf)) {
        throw new Error('FormComponentFactoryProvider: AnyOf should be handled in the scope of the ObjectField');
      }

      return DisabledField;
    },
    [customFieldsFactory],
  );

  return <FormComponentFactoryContext.Provider value={factory}>{children}</FormComponentFactoryContext.Provider>;
};

/* Name of the properties that should load TextAreaField */
const TEXT_AREA_PROPERTY_NAMES = ['Expression', 'Description', 'Query', 'Script'];
