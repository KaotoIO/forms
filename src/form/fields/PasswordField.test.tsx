import { act, fireEvent, render, waitFor, within } from '@testing-library/react';
import { ModelContextProvider } from '../providers/ModelProvider';
import { SchemaProvider } from '../providers/SchemaProvider';
import { SuggestionContext } from '../providers/SuggestionRegistryProvider';
import { ROOT_PATH } from '../utils';
import { PasswordField } from './PasswordField';

describe('PasswordField', () => {
  const mockSuggestionProvider = {
    id: 'test-provider',
    appliesTo: jest.fn().mockReturnValue(true),
    getSuggestions: jest.fn().mockResolvedValue([
      { value: 'test-suggestion-1', description: 'First test suggestion' },
      { value: 'test-suggestion-2', description: 'Second test suggestion' },
    ]),
  };

  const getProvidersMock = jest.fn().mockReturnValue([mockSuggestionProvider]);

  const renderWithSuggestions = (children: React.ReactNode) => {
    return render(
      <SuggestionContext.Provider value={{ getProviders: getProvidersMock }}>{children}</SuggestionContext.Provider>,
    );
  };

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should render', () => {
    const { container } = render(
      <ModelContextProvider model="Value" onPropertyChange={jest.fn()}>
        <PasswordField propName={ROOT_PATH} />
      </ModelContextProvider>,
    );

    expect(container).toMatchSnapshot();
  });

  it('should set the appropriate placeholder', () => {
    const wrapper = render(
      <ModelContextProvider model={undefined} onPropertyChange={jest.fn()}>
        <SchemaProvider schema={{ type: 'string', default: 'Default Value' }}>
          <PasswordField propName={ROOT_PATH} />
        </SchemaProvider>
      </ModelContextProvider>,
    );

    const input = wrapper.getByRole('textbox');
    expect(input).toHaveAttribute('placeholder', 'Default Value');
  });

  it('should notify when the value changes', () => {
    const onPropertyChangeSpy = jest.fn();

    const wrapper = render(
      <ModelContextProvider model="Value" onPropertyChange={onPropertyChangeSpy}>
        <PasswordField propName={ROOT_PATH} />
      </ModelContextProvider>,
    );

    const input = wrapper.getByRole('textbox');
    act(() => {
      fireEvent.change(input, { target: { value: 'New Value' } });
    });

    expect(onPropertyChangeSpy).toHaveBeenCalledTimes(1);
    expect(onPropertyChangeSpy).toHaveBeenCalledWith(ROOT_PATH, 'New Value');
  });

  it('should clear the input when using the clear button', async () => {
    const onPropertyChangeSpy = jest.fn();

    const wrapper = render(
      <ModelContextProvider model="Value" onPropertyChange={onPropertyChangeSpy}>
        <PasswordField propName={ROOT_PATH} />
      </ModelContextProvider>,
    );

    const fieldActions = wrapper.getByTestId(`${ROOT_PATH}__field-actions`);
    act(() => {
      fireEvent.click(fieldActions);
    });

    const clearButton = await wrapper.findByRole('menuitem', { name: /clear/i });
    act(() => {
      fireEvent.click(clearButton);
    });

    expect(onPropertyChangeSpy).toHaveBeenCalledTimes(1);
    expect(onPropertyChangeSpy).toHaveBeenCalledWith(ROOT_PATH, undefined);
  });

  it('should use onRemoveProps callback if specified when using the clear button', async () => {
    const onPropertyChangeSpy = jest.fn();
    const onRemoveSpy = jest.fn();

    const wrapper = render(
      <ModelContextProvider model="Value" onPropertyChange={onPropertyChangeSpy}>
        <PasswordField propName={ROOT_PATH} onRemove={onRemoveSpy} />
      </ModelContextProvider>,
    );

    const fieldActions = wrapper.getByTestId(`${ROOT_PATH}__field-actions`);
    act(() => {
      fireEvent.click(fieldActions);
    });

    const clearButton = await wrapper.findByRole('menuitem', { name: /remove/i });
    act(() => {
      fireEvent.click(clearButton);
    });

    expect(onPropertyChangeSpy).not.toHaveBeenCalled();
    expect(onRemoveSpy).toHaveBeenCalledTimes(1);
    expect(onRemoveSpy).toHaveBeenCalledWith(ROOT_PATH);
  });

  it('should show errors if available for its property path', () => {
    const onPropertyChangeSpy = jest.fn();

    const wrapper = render(
      <ModelContextProvider
        model="Value"
        errors={{ [ROOT_PATH]: ['error message'] }}
        onPropertyChange={onPropertyChangeSpy}
      >
        <PasswordField propName={ROOT_PATH} />
      </ModelContextProvider>,
    );

    const errorMessage = wrapper.getByText('error message');
    expect(errorMessage).toBeInTheDocument();
  });

  it('should hide password by default', () => {
    const wrapper = render(
      <ModelContextProvider model="Value" onPropertyChange={jest.fn()}>
        <PasswordField propName={ROOT_PATH} />
      </ModelContextProvider>,
    );

    const input = wrapper.getByRole('textbox');

    expect(input.getAttribute('type')).toBe('password');
  });

  it('should toggle the password visibility', () => {
    const wrapper = render(
      <ModelContextProvider model="Value" onPropertyChange={jest.fn()}>
        <PasswordField propName={ROOT_PATH} />
      </ModelContextProvider>,
    );

    const toggleButton = wrapper.getByTestId(`${ROOT_PATH}__toggle-visibility`);
    act(() => {
      fireEvent.click(toggleButton);
    });

    const input = wrapper.getByRole('textbox');

    expect(input.getAttribute('type')).toBe('text');
  });

  it('wraps value with RAW when Raw button is clicked', async () => {
    const onPropertyChangeSpy = jest.fn();

    const wrapper = render(
      <ModelContextProvider model="SecretPassword" onPropertyChange={onPropertyChangeSpy}>
        <PasswordField propName={ROOT_PATH} />
      </ModelContextProvider>,
    );

    const fieldActions = wrapper.getByTestId(`${ROOT_PATH}__field-actions`);
    act(() => {
      fireEvent.click(fieldActions);
    });

    const rawItem = await wrapper.findByRole('menuitem', { name: /raw/i });
    act(() => {
      fireEvent.click(rawItem);
    });

    expect(onPropertyChangeSpy).toHaveBeenCalledTimes(1);
    expect(onPropertyChangeSpy).toHaveBeenCalledWith(ROOT_PATH, 'RAW(SecretPassword)');
  });

  it('unwraps value from RAW when already wrapped', async () => {
    const onPropertyChangeSpy = jest.fn();

    const wrapper = render(
      <ModelContextProvider model="RAW(SecretPassword)" onPropertyChange={onPropertyChangeSpy}>
        <PasswordField propName={ROOT_PATH} />
      </ModelContextProvider>,
    );

    const fieldActions = wrapper.getByTestId(`${ROOT_PATH}__field-actions`);
    act(() => {
      fireEvent.click(fieldActions);
    });

    const rawItem = await wrapper.findByRole('menuitem', { name: /raw/i });
    act(() => {
      fireEvent.click(rawItem);
    });

    expect(onPropertyChangeSpy).toHaveBeenCalledTimes(1);
    expect(onPropertyChangeSpy).toHaveBeenCalledWith(ROOT_PATH, 'SecretPassword');
  });

  it('should show suggestions when Ctrl+Space is pressed', async () => {
    const onPropertyChangeSpy = jest.fn();

    const wrapper = renderWithSuggestions(
      <ModelContextProvider model="" onPropertyChange={onPropertyChangeSpy}>
        <PasswordField propName={ROOT_PATH} />
      </ModelContextProvider>,
    );

    await act(async () => {
      const input = wrapper.getByRole('textbox');
      input.focus();
      fireEvent.keyDown(input, { code: 'Space', ctrlKey: true });
    });

    await waitFor(() => {
      expect(wrapper.getByTestId('suggestions-menu')).toBeInTheDocument();
    });

    // Check if suggestions are rendered
    expect(wrapper.getByText('First test suggestion')).toBeInTheDocument();
    expect(wrapper.getByText('Second test suggestion')).toBeInTheDocument();
  });

  it('should apply suggestion when clicked', async () => {
    const onPropertyChangeSpy = jest.fn();

    const wrapper = renderWithSuggestions(
      <ModelContextProvider model="" onPropertyChange={onPropertyChangeSpy}>
        <PasswordField propName={ROOT_PATH} />
      </ModelContextProvider>,
    );

    await act(async () => {
      const input = wrapper.getByRole('textbox');
      input.focus();
      fireEvent.keyDown(input, { code: 'Space', ctrlKey: true });
    });

    const suggestionMenu = await wrapper.findByRole('menu');
    expect(suggestionMenu).toBeInTheDocument();

    const firstSuggestion = within(suggestionMenu).getByText('First test suggestion');
    await act(async () => {
      fireEvent.click(firstSuggestion);
    });

    expect(onPropertyChangeSpy).toHaveBeenCalledWith(ROOT_PATH, 'test-suggestion-1');
  });
});
