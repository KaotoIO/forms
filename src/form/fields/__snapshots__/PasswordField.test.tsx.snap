// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PasswordField should render 1`] = `
<div>
  <div
    class="pf-v6-c-form__group"
    data-testid="#__field-wrapper"
  >
    <div
      class="pf-v6-c-form__group-label"
    >
      <label
        class="pf-v6-c-form__label"
        for="#"
      >
        <span
          class="pf-v6-c-form__label-text"
        >
          #
        </span>
      </label>
        
      <span
        class="pf-v6-c-form__group-label-help"
      >
        <div
          style="display: contents;"
        >
          <span
            aria-label="More info for # field"
            class="pf-v6-c-button pf-m-plain pf-m-no-padding"
            data-ouia-component-id="OUIA-Generated-Button-plain-1"
            data-ouia-component-type="PF6/Button"
            data-ouia-safe="true"
            role="button"
            tabindex="0"
            type="button"
          >
            <span
              class="pf-v6-c-button__icon"
            >
              <svg
                aria-hidden="true"
                class="pf-v6-svg"
                fill="currentColor"
                height="1em"
                role="img"
                viewBox="0 0 1024 1024"
                width="1em"
              >
                <path
                  d="M521.3,576 C627.5,576 713.7,502 713.7,413.7 C713.7,325.4 627.6,253.6 521.3,253.6 C366,253.6 334.5,337.7 329.2,407.2 C329.2,414.3 335.2,416 343.5,416 L445,416 C450.5,416 458,415.5 460.8,406.5 C460.8,362.6 582.9,357.1 582.9,413.6 C582.9,441.9 556.2,470.9 521.3,473 C486.4,475.1 447.3,479.8 447.3,521.7 L447.3,553.8 C447.3,570.8 456.1,576 472,576 C487.9,576 521.3,576 521.3,576 M575.3,751.3 L575.3,655.3 C575.313862,651.055109 573.620137,646.982962 570.6,644 C567.638831,640.947672 563.552355,639.247987 559.3,639.29884 L463.3,639.29884 C459.055109,639.286138 454.982962,640.979863 452,644 C448.947672,646.961169 447.247987,651.047645 447.29884,655.3 L447.29884,751.3 C447.286138,755.544891 448.979863,759.617038 452,762.6 C454.961169,765.652328 459.047645,767.352013 463.3,767.30116 L559.3,767.30116 C563.544891,767.313862 567.617038,765.620137 570.6,762.6 C573.659349,759.643612 575.360354,755.553963 575.3,751.3 M512,896 C300.2,896 128,723.9 128,512 C128,300.3 300.2,128 512,128 C723.8,128 896,300.2 896,512 C896,723.8 723.7,896 512,896 M512.1,0 C229.7,0 0,229.8 0,512 C0,794.2 229.8,1024 512.1,1024 C794.4,1024 1024,794.3 1024,512 C1024,229.7 794.4,0 512.1,0"
                />
              </svg>
            </span>
          </span>
        </div>
      </span>
    </div>
    <div
      class="pf-v6-c-form__group-control"
    >
      <div
        class="pf-v6-c-text-input-group"
      >
        <div
          aria-describedby="#-popover"
          class="pf-v6-c-text-input-group__main"
          id="#"
        >
          <span
            class="pf-v6-c-text-input-group__text"
          >
            <input
              aria-label="Type to filter"
              class="pf-v6-c-text-input-group__text-input"
              name="#"
              role="textbox"
              type="password"
              value="Value"
            />
          </span>
        </div>
        <div
          class="pf-v6-c-text-input-group__utilities"
        >
          <button
            aria-label="Show # value"
            class="pf-v6-c-button pf-m-plain"
            data-ouia-component-id="OUIA-Generated-Button-plain-2"
            data-ouia-component-type="PF6/Button"
            data-ouia-safe="true"
            data-testid="#__toggle-visibility"
            type="button"
          >
            <span
              class="pf-v6-c-button__text"
            >
              <svg
                aria-hidden="true"
                class="pf-v6-svg"
                fill="currentColor"
                height="1em"
                role="img"
                viewBox="0 0 576 512"
                width="1em"
              >
                <path
                  d="M572.52 241.4C518.29 135.59 410.93 64 288 64S57.68 135.64 3.48 241.41a32.35 32.35 0 0 0 0 29.19C57.71 376.41 165.07 448 288 448s230.32-71.64 284.52-177.41a32.35 32.35 0 0 0 0-29.19zM288 400a144 144 0 1 1 144-144 143.93 143.93 0 0 1-144 144zm0-240a95.31 95.31 0 0 0-25.31 3.79 47.85 47.85 0 0 1-66.9 66.9A95.78 95.78 0 1 0 288 160z"
                />
              </svg>
            </span>
          </button>
          <button
            aria-expanded="false"
            aria-label="#__field-actions"
            class="pf-v6-c-menu-toggle pf-m-plain"
            data-ouia-component-id="OUIA-Generated-MenuToggle-plain-1"
            data-ouia-component-type="PF6/MenuToggle"
            data-ouia-safe="true"
            data-testid="#__field-actions"
            type="button"
          >
            <span
              class="pf-v6-c-menu-toggle__icon"
            >
              <svg
                aria-hidden="true"
                class="pf-v6-svg"
                fill="currentColor"
                height="1em"
                role="img"
                viewBox="0 0 192 512"
                width="1em"
              >
                <path
                  d="M96 184c39.8 0 72 32.2 72 72s-32.2 72-72 72-72-32.2-72-72 32.2-72 72-72zM24 80c0 39.8 32.2 72 72 72s72-32.2 72-72S135.8 8 96 8 24 40.2 24 80zm0 352c0 39.8 32.2 72 72 72s72-32.2 72-72-32.2-72-72-72-72 32.2-72 72z"
                />
              </svg>
            </span>
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
